lib/cdo/active_job_backend.rb

```
@@ -0,0 +1,240 @@
+
+      # We delete ALL existing delayed_job pid_files in pid_dir (dashboard/tmp/pids)
+      # before we even start in order to work around a horrible bug in delayed_job:
+      #
+      # If delayed_job.2.pid doesn't exist, but delayed_job.22.pid does exist,
+      # delayed_job will think delayed_job.22.pid is delayed_job.2 and therefore
+      # already running, so when you try to start delayed_job.2 it will refuse
+      # to start and error out with:
+      #     ERROR: there is already one or more instance(s) of the program running
+      #
+      # Deleting all the pid_files before killing the processes isn't as bad as it sounds, since:
+      # 1. We already fetched pids, and we're gonna kill them all anyway
+      # 2. Even if we crash, we actually consider the union of pid files and ps output
+      #    so losing the pid files is completely recoverable, we'll still reap those processes
+      #    next time.
+      delete_pid_files(pids, pid_file_hash)
+
+      # Stop and Start workers in equal-sized rolling batches to prevent downtime
+      n_workers_started = 0
+      pids.each_slice(n_workers_to_restart_per_batch) do |pids_in_batch|
+        # Stop pre-existing delayed_job workers in this batch
+        stop_workers(pids_in_batch, pid_file_hash)
+
+        # Start (up to) an equal number of replacement workers
+        n_workers = (n_workers_to_start - n_workers_started).clamp(0, pids_in_batch.size)
+        n_workers_started += start_n_workers(n_workers, initial_worker_index: n_workers_started) if n_workers > 0
+      end
+
+      # Start any remaining workers (=we're starting more workers than previously existed)
+      n_workers = n_workers_to_start - n_workers_started
+      n_workers_started += start_n_workers(n_workers, initial_worker_index: n_workers_started) if n_workers > 0
+
+      check_worker_status(n_workers_to_start)
+    end
+
+    def self.check_worker_status(n_workers_to_start)
```

@davidsbailey fyi this function run post-deploy already checks that number of workers deployed equals number intended to start, however, I don't fail the deploy in this condition (I do fail it if no workers were started).

I'd be interested in your thoughts on whether we should count n_workers_to_start != n_workers_running as a failed deploy. My concern is that when deploying 150 workers, the number is high enough that one or a couple might not occasionally start for ??? reasons. If we were deploying 150k workers I would pretty much guarantee that "fail all if any fail" is a bad strategy that will lead to lots of unnecessarily failed deploys.

I like the idea of adding an "ensure all pids older than deploy start" ps check.

---

Maybe we should have fuzziness on fail: if more than "5%" of workers fail, fail the deploy. If less than 5% (but not 0) workers fail, warn. If ANY workers in `ps` are older than the start of our worker restart, fail deploy.

---

I've added:
- a "fail hard" check for "no workers in ps older than the deploy start time"
- a "fail hard" check for number of running workers against an `ACCEPTABLE_WORKER_FAILURE_RATE` > 2%
  - produces a warning if any workers failed but failure to start rate is <2%

---

sorry I missed this code the first time. sounds great!

---

I /really/ like the "make sure no old processes exist" check, thanks for thinking of that